frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+      }
+    },
+    'gi.contracts/chatroom/leave': {
+      validate: objectOf({
+        username: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { username } = data
+        if (state.users[username] && !state.users[username].departedDate) {
+          if (!sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+            Vue.set(state.users[username], 'departedDate', meta.createdDate)
+          }
+
+          const notificationType = username === meta.username ? MESSAGE_NOTIFICATIONS.LEAVE_MEMBER : MESSAGE_NOTIFICATIONS.KICK_MEMBER
+          const notificationData = createNotificationData(
+            notificationType,
+            notificationType === MESSAGE_NOTIFICATIONS.KICK_MEMBER ? { username } : {})
+          const newMessage = createMessage({ meta, hash, data: notificationData, state })
+          Vue.set(state.messages, [state.messages.length], newMessage)
+          return
+        }
+        console.log(`chatroom Leave: ${username} is not a member of this chatroom #${state.name}`)
+      },
+      sideEffect ({ data, hash, contractID }, { state }) {
+        if (sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+          return
+        }
+        const rootState = sbp('state/vuex/state')
+        if (data.username === rootState.loggedIn.username) {
+          leaveChatRoom({ contractID })
+        }
+        emitMessageEvents({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/delete': {
+      validate: (data, { state, getters, meta }) => {
+        if (state.attributes.creator !== meta.username) {
+          throw new TypeError(L('Only the channel creator can delete channel.'))
+        }
+      },
+      process ({ data, meta }, { state, rootState }) {
+        Vue.set(state.attributes, 'deletedDate', meta.createdDate)
+        for (const username in state.users) {
+          Vue.set(state.users[username], 'departedDate', meta.createdDate)
+        }
+      },
+      sideEffect ({ meta, contractID }, { state }) {
+        if (state.attributes.creator === meta.username) { // Not sure this condition is necessary
+          if (sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+            return
+          }
+          leaveChatRoom({ contractID })
+        }
+      }
+    },
+    'gi.contracts/chatroom/addMessage': {
+      validate: objectMaybeOf({
+        type: unionOf(...Object.values(MESSAGE_TYPES).map(v => literalOf(v))),
+        text: string,
+        notification: objectMaybeOf({
+          type: unionOf(...Object.values(MESSAGE_NOTIFICATIONS).map(v => literalOf(v))),
+          params: mapOf(string, string) // { username }
+        }),
+        replyingMessage: objectOf({
+          id: string, // scroll to the original message and highlight
+          username: string, // display
+          text: string, // display
+          time: string // to search easily
+        }),
+        onlyVisibleTo: arrayOf(string)
+      }),
```

Comment:
This looks like it might be a DRY violation? You already defined `messageType` above...

Reply:
Correct. I guess I forgot to use it.